### PR TITLE
Add cancel and delete to IonQ job

### PR DESCRIPTION
**Pull request by dabacon**
Also:
* factors out ``IonQExceptions`
* checks that jobs are not canceled or deleted when jobs are queried for fields that no longer exists when job enteres those states.

### CODE REVIEW

File: cirq/ionq/ionq_client.py [hunk 1]
```diff
@@ -226,10 +208,10 @@ def get_job(self, job_id: str):
         """Get the job from the IonQ API.
 
         Args:
-            job_id: The UUID of the job (returned when the job is created).
+            job_id: The UUID of the job (returned when the job was created).
 
         Returns:
-            A `cirq.ionq.IonQJob` corresponding to the job.
+            The json body of the response a dict.
```
Comment (95-martin-orion) [inline cirq/ionq/ionq_client.py:214]: "a dict" -> "as a dict"

File: cirq/ionq/ionq_client_test.py [hunk 2]
```diff
@@ -317,3 +317,38 @@ def test_ionq_client_get_job_retry(mock_get):
                                           default_target='simulator')
     _ = client.get_job('job_id')
     assert mock_get.call_count == 2
+
+
+@mock.patch('requests.put')
+def test_ionq_client_cancel_job(mock_put):
```
Comment (95-martin-orion) [inline cirq/ionq/ionq_client_test.py:323]: Should this and `delete_job` also have failure tests, or is the coverage provided by the `get` and `create` tests sufficient?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Looks good overall, just a couple points to check on.

File: cirq/ionq/ionq_client.py (hunk 1 above)
Comment (dabacon) [reply cirq/ionq/ionq_client.py:214]: fixed

File: cirq/ionq/ionq_client_test.py (hunk 2 above)
Comment (dabacon) [reply cirq/ionq/ionq_client_test.py:323]: Probably but better form to test these.  I've added tests.  What's the new acronym for the opposite of dry in tests?

**Comment by dabacon** [comment]
PTAL

**Comment by dabacon** [comment]
Thanks for the quick review :)

File: cirq/ionq/ionq_client_test.py (hunk 2 above)
Comment (95-martin-orion) [reply cirq/ionq/ionq_client_test.py:323]: Internet tells me it's WET (for "Write Everything Twice") 😝 

I have a feeling these could be parameterized to DRY things up, but I'm not going to block the merge on that. LGTM

**Review by 95-martin-orion — APPROVED** [review]

**Comment by dabacon** [comment]
Part of #3479

**Pull request merged**